### DISCUSSION TITLE
PyMoo implementation of FDA Dynamic multi objective test problems

### DISCUSSION THREAD

**Discussion by bandundu** [anyoptimization/pymoo#286] [Ideas]
Hello dear community,

for my master thesis, I plan on testing / implementing several dynamic multi objective evolutionary algorithms (DMOEAs). 
Therefore, I want to use pymoo as it turned out to be a great tool for previous analyses I did.

Sadly, pymoo does not provide any dynamic multi objective problems (DMOPs), such as the [FDA problem class](https://www.researchgate.net/profile/Kalyan-Deb/publication/3418802_Dynamic_Multiobjective_Optimization_Problems_Test_Cases_Approximations_and_Applications/links/02e7e51938c327abb7000000/Dynamic-Multiobjective-Optimization-Problems-Test-Cases-Approximations-and-Applications.pdf?_sg%5B0%5D=SaddIyTcW-aOLnpX2n0nWDjLGoLn47KXqNRmjYtJ-TP91sb6CO6WWWe-yIy-vZ0brUV1-_vFKHpNsRFDWtGQjg.6vM0FdMeihSzvrB1SuO30ZiZBmr8en34xHo79XqMSoSuxFp45R_VOmsGQhQQ_3JMNFFrNna9zvR-jXTluP4XPQ&_sg%5B1%5D=GsFZxdh8qrumd8U2hr0t0XQFrgk66PdV2B2qsxEHxbskjKtOseLHWU4zE_rxEo9TxStXMpTVCn_9aQ1ZUNaksfMw9sL_Dyd1TpIQj9ch6DsE.6vM0FdMeihSzvrB1SuO30ZiZBmr8en34xHo79XqMSoSuxFp45R_VOmsGQhQQ_3JMNFFrNna9zvR-jXTluP4XPQ&_iepl=). 

I am aware that  [jmetalpy](https://github.com/jMetal/jMetalPy) provides this functionality, but I prefer pymoo as a framework since it easily enables stats reviews as well as [biased initialization](https://pymoo.org/customization/initialization.html)

Now before I dive into implementing the FDA Problem class in pymoo I wanted to make sure, whether this is feasible in the framework or if I might be headed in the wrong direction. 

My current plan is doing something like this:
```
class FDA4(Problem):
    def __init__(self):
       ...
       self.time = 0
       
    def update(self):
        self.time += 1

    def _evaluate(self, x, out, *args, **kwargs):
        out["F"] = np.sum((x - 0.5) ** 2 * self.time, axis=1)

```

```
# Define Algorithm
algorithm = FDA4()

# Minimize with NSGA-II
res = minimize(problem, algorithm, ("n_gen", 10), seed=seed, verbose=True)

POF = res.F
POS = res.X

# Update Problem
problem.update()

# Predict a preferable initial population

# Minimize with NSGA-II with biased initialization
res = minimize(problem, algorithm, ("n_gen", 10), seed=seed, verbose=True)
```

Would there be any issues by implementing this in that kind of way? Thank you very much for any response, I'm still new to the field so apologies for imprecise statements.

**Comment by blankjul** [comment]
Please see my post on GitHub issues. I am more than happy to have a chat about it. I think having a callback that iterates on the problem is a cleaner solution. 
In general, the algorithm needs to figure out: ohhh, the problem has changed and re-evaluate existing solutions. It makes more sense to design an algorithm directly instead of calling `minimize` multiple times.

Again, I am more than happy to have a chat if you are interested working on dynamic problems in pymoo.
